### PR TITLE
Refactor Survey Translation

### PR DESCRIPTION
**Pull request by iwootten**
**Motivation**

Current template extraction and translation is complicated and has limited test coverage. Common behaviour of both extraction and translation is duplicated.

Some strings are currently not translated and it is difficult to debug and understand why. In addition some features are covered by existing libraries, such as babel.

**Changes**

Search schemas for known strings to extract and translate using the same common logic, based on [jsonpointer](https://python-json-pointer.readthedocs.io/en/latest/). Use babel to read and write to po/pot files and remove all csv behaviour to be replaced with if ever required again [translate-toolkit](http://docs.translatehouse.org/projects/translate-toolkit/en/latest/commands/csv2po.html).

### CODE REVIEW

**Comment by jonnyshaw89** [comment]
When i run `template_extractor` against the `census_household` schema in runner i get some wrapping on the `msgid` eg...
```
msgid ""
"What are the names of everyone who live at "
"<em>{{metadata['display_address']}}</em>?"
msgstr ""
```

File: README.md [hunk 1]
```diff
@@ -16,27 +16,28 @@ pipenv install --dev
...
-This will output the translatable text to an Excel (.xlsx) file.
+This will output the translatable text to an POT file.
+
+
+After the text has been translated, create a new translated survey with:
 
-After the text has been translated, create a new translated survey with
 ```
-pipenv run ./scripts/run_translate_survey.sh <json_file> <translations_file> [output_directory]
+pipenv run python -m cli.translate_survey <schema_file> <output_directory>
```
Comment (jonnyshaw89) [inline README.md]: this is missing the `TRANSLATION_PATH`?

`pipenv run python -m cli.translate_survey <schema_file> <translation_path> <output_directory>`

**Comment by iwootten** [comment]
> When i run `template_extractor` against the `census_household` schema in runner i get some wrapping on the `msgid` eg...
> 
> ```
> msgid ""
> "What are the names of everyone who live at "
> "<em>{{metadata['display_address']}}</em>?"
> msgstr ""
> ```
> Or is this just the PO file format?

Yeah, that's how babel writes it's po files.

File: README.md (hunk 1 above)
Comment (iwootten) [reply README.md]: So it is, updated.

File: app/schema_translation.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                found = True
+                if message.auto_comments:
+                    first_comment = message.auto_comments[0]
+                    answer_id = first_comment.split(":")[1].strip()
+                    found = answer_id == message_to_translate['answer_id']
+
+                if found:
+                    return dumb_to_smart_quotes(message.string)
+
+    def translate(self, messages):
```
Comment (ajmaddaford) [inline app/schema_translation.py]: I can only see this method being used in tests, or have I missed something?

File: app/schema_translation.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            pofile.write_po(translation_file, self.catalog)
+
+    def translate_message(self, message_to_translate):
+        for message in self.catalog:
+            if message.id and dumb_to_smart_quotes(message.id) == dumb_to_smart_quotes(message_to_translate['original']):
+                found = True
+                if message.auto_comments:
+                    first_comment = message.auto_comments[0]
+                    answer_id = first_comment.split(":")[1].strip()
+                    found = answer_id == message_to_translate['answer_id']
```
[the remark this answers is not in the document]
Comment (iwootten) [reply app/schema_translation.py]: Makes sense, updated.

File: app/survey_schema.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                missing_translations += 1
+                print("Missing translation at {}: '{}'".format(pointer, pointer_contents))
+
+        for pointer in self.context_pointers:
+            pointer_contents = resolve_pointer(self.schema, pointer)
+            parent_answer_id = self.get_parent_id(pointer)
+            translation = schema_translation.translate_message({
+                'original': pointer_contents,
+                'answer_id': parent_answer_id
+            })
```
[the remark this answers is not in the document]
Comment (iwootten) [reply app/survey_schema.py]: Ok, I've updated this.

File: app/schema_translation.py (hunk 2 above)
Comment (iwootten) [reply app/schema_translation.py]: Yes, it was. Removed it now.

**Comment by ajmaddaford** [comment]
I've just tried running this on test_summary.json and there seems to be an issue with radio options with detail answers missing the question text in the message context:
```
#. answer-id: radio-answer
msgctxt "Answer for: "
msgid "Other"
msgstr ""

#. answer-id: other-answer-mandatory
msgctxt "Answer for: "
msgid "Please specify other"
msgstr ""
```

**Comment by iwootten** [comment]
@ajmaddaford Looks like that is happening because there isn't a title defined on the question..instead it appears at the block level.

**Comment by ajmaddaford** [comment]
@iwootten Fair enough then, a question should always have a title.

**Pull request merged**